readium/navigator/src/main/java/org/readium/r2/navigator/media3/tts/TtsNavigator.kt

```diff
@@ -0,0 +1,265 @@
+            E : TtsEngine.Error, V : TtsEngine.Voice> invoke(
+            application: Application,
+            publication: Publication,
+            ttsEngineProvider: TtsEngineProvider<S, P, *, E, V>,
+            tokenizerFactory: (defaultLanguage: Language?) -> ContentTokenizer,
+            metadataProvider: MediaMetadataProvider,
+            listener: Listener,
+            initialPreferences: P? = null,
+            initialLocator: Locator? = null,
+        ): TtsNavigator<S, P, E, V>? {
+
+            if (publication.findService(ContentService::class) == null) {
+                return null
+            }
+
+            val actualInitialPreferences = initialPreferences
+                ?: ttsEngineProvider.createEmptyPreferences()
+
+            val contentIterator =
+                TtsContentIterator(publication, tokenizerFactory, initialLocator)
+
+            val ttsEngine =
+                ttsEngineProvider.createEngine(publication, actualInitialPreferences)
+                    ?: return null
+
+            val metadataFactory =
+                metadataProvider.createMetadataFactory(publication)
+
+            val playlistMetadata =
+                metadataFactory.publicationMetadata()
+
+            val mediaItems =
+                publication.readingOrder.indices.map { index ->
+                    val metadata = metadataFactory.resourceMetadata(index)
+                    MediaItem.Builder()
+                        .setMediaMetadata(metadata)
+                        .build()
+                }
+
+            val ttsPlayer =
+                TtsPlayer(ttsEngine, contentIterator, actualInitialPreferences)
+                    ?: return null
+
+            val coroutineScope =
+                MainScope()
+
+            val playbackParameters =
+                ttsPlayer.settings.mapStateIn(coroutineScope) {
+                    ttsEngineProvider.getPlaybackParameters(it)
+                }
+
+            val onSetPlaybackParameters = { parameters: PlaybackParameters ->
+                val newPreferences = ttsEngineProvider.updatePlaybackParameters(
+                    ttsPlayer.lastPreferences,
+                    parameters
+                )
+                ttsPlayer.submitPreferences(newPreferences)
+            }
+
+            val sessionAdapter =
+                TtsSessionAdapter(
+                    application,
+                    ttsPlayer,
+                    playlistMetadata,
+                    mediaItems,
+                    listener::onStopRequested,
+                    playbackParameters,
+                    onSetPlaybackParameters,
+                    ttsEngineProvider::mapEngineError
+                )
+
+            return TtsNavigator(coroutineScope, publication, ttsPlayer, sessionAdapter)
+        }
+    }
+
+    interface Listener {
+
+        fun onStopRequested()
+
+        fun onMissingLanguageData(language: Language)
```

Comment:
I've implemented the same solution as in `PdfEngineProvider` : a custom listener. It's not used yet in the testapp as it would be a bit tricky because of its structure.